### ISSUE TITLE
Switching from Exrm to Distillery : erratic behaviour

### ISSUE THREAD

**Issue by Moxide** [boldpoker/edeliver#132]
Hello,
I'm trying to use Edeliver with Distillery and I'm facing an issue. (Elixir 1.3.2, Erlang 19)
When I run **mix edeliver build release**, everything runs as usual until "Generating release" :
```
-----> Generating release
using mix to generate release
Changed release version from 0.0.1 to 0.0.1+distillery-60a45a7-20160806-prod
Building release with MIX_ENV=prod.
** (FunctionClauseError) no function clause matching in Mix.env/1
    (mix) lib/mix.ex:211: Mix.env("prod")
    lib/exrm/utils/utils.ex:19: ReleaseManager.Utils.with_env/2
    lib/mix/tasks/release.ex:164: Mix.Tasks.Release.generate_sys_config/1
    lib/mix/tasks/release.ex:74: Mix.Tasks.Release.do_run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3

A remote command failed on:

  mox@buildserver

Output of the command is shown above and the command executed
on that host is printed below for debugging purposes:

FAILED with exit status 1:

    [ -f ~/.profile ] && source ~/.profile
    set -e
    cd /tmp/erlang/my-app/builds/App
    if [ "mix" = "rebar" ]; then
      echo "using rebar to generate release"
      ./rebar   -f generate
    elif [ "mix" = "relx" ]; then
      echo "using relx to generate release"
      ./relx release
    elif [ "mix" = "mix" ]; then
      echo "using mix to generate release"
      MIX_ENV="prod" LINK_SYS_CONFIG="" LINK_VM_ARGS="" APP="App" AUTO_VERSION="git-branch+git-revision+build-date+mix-env" BRANCH="distillery" SKIP_RELUP_MODIFICATIONS="" RELUP_MODIFICATION_MODULE="" mix do release.version  --verbose,  release --verbose --env="prod" --name="App"
    fi
```
When I ssh to the build server and run **MIX_ENV=prod mix help release**, the release task is the Exrm's one and not the Distillery's, hence the error.

In **mix.lock**, both Distillery and Exrm dependencies appears. In Edeliver line, Distillery is defined as optional and Exrm is not.

If I switch this and rerun **MIX_ENV=prod mix help release**, the Distillery's task shows up. And if I run the above offending line 
`MIX_ENV="prod" LINK_SYS_CONFIG="" LINK_VM_ARGS="" APP="App" AUTO_VERSION="git-branch+git-revision+build-date+mix-env" BRANCH="distillery" SKIP_RELUP_MODIFICATIONS="" RELUP_MODIFICATION_MODULE="" mix do release.version  --verbose,  release --verbose --env="prod" --name="App"`
It succeeds.

Since Exrm and Distillery both declare a task named "release", how Elixir is prioritizing one on the other ?
Thanks in advance !

**Comment by bharendt** [comment]
Hello, 
if edeliver detects that distillery is used as release tool, actually `distillery` [should be used as dependency](https://github.com/boldpoker/edeliver/blob/c2db9cb4c2230d7981ec9e0867fcc815e5d79023/mix.exs#L37-L41) and `exrm` exrm should be marked as `optional` depedency. 
Edeliver also [adjusts it's build path](https://github.com/boldpoker/edeliver/blob/c2db9cb4c2230d7981ec9e0867fcc815e5d79023/mix.exs#L67) because it provides release plugins for both, `exrm` and `destillery`, which have compile-time dependencies to the particular build tool.

The __detection__ that `distillery` is used as build to depends on two things: that `distillery` [is __added as dependency to your project__](https://github.com/boldpoker/edeliver/blob/c2db9cb4c2230d7981ec9e0867fcc815e5d79023/mix.exs#L79), and that a [distillery config extsts](https://github.com/boldpoker/edeliver/blob/b76fa6fbd0a28653e911596d9cb9c570b184ef2a/libexec/app_config#L97) at `rel/config.exs`.

__When switching from `exrm` to `distillery` as build tool, it is required to remove the `exrm` dependency from__ both, the __`mix.exs`__ file __and__ from the __`mix.lock`__ file, because they both provide two modules with the same name: `Mix.Tasks.Release` and `Mix.Tasks.Release.Clean` as you already mentioned above.

In short: switching from `exrm` to `distillery` requires the following steps:

 - add `distillery` as dependency to the the mix file
 - remove `exrm` as dependency from the `mix.exs` and `mix.lock` file
 - create a `distillery` config at `rel/config.exs`, maybe by using `mix release.init`
 - use `edeliver` version >= `1.3.0` for deployment

**Comment by Moxide** [comment]
Thanks for the fast answer !

> In short: switching from exrm to distillery requires the following steps:
> 
>     add distillery as dependency to the the mix file
>     remove exrm as dependency from the mix.exs and mix.lock file
>     create a distillery config at rel/config.exs, maybe by using mix release.init
>     use edeliver version >= 1.3.0 for deployment
> 

This is exactly what I did, apart that I added `USING_DISTILLERY=true` in `.deliver/config`.
Trying to narrow down the issue, I did this :
```
rm -Rf _build
rm -Rf deps
rm mix.lock
MIX_ENV=prod mix deps.get
```
`mix.lock` contains **both** Exrm and Distillery dependencies, and in Edeliver line, Distillery is optional and Exrm is not.

It seems that the 'optional' parameter is ignored and/or the distillery detection routine is not working is my case.
I'll investigate further and report back.

**Comment by bharendt** [comment]
Setting `USING_DISTILLERY=true` is a good idea in that case. But if there is still `exrm` used as dependency, it might be that another library depends on it. You could maybe check that with a command like `grep -ne 'exrm' deps/*/mix.exs`.

**Comment by Moxide** [comment]
Did that too, no other library uses exrm :-(
I'll give it a try tomorrow again...

Thank you very much for your help !

**Comment by Moxide** [comment]
Just tried the following code in edeliver's `mix.exs` (removing the 2 "optional" lines) :
```
  if project_uses_distillery? do
      [
        {:distillery, ">= 0.8.0", warn_missing: false},
        # {:exrm, ">= 0.16.0", optional: true, warn_missing: false},
      ]
    else
      [
        {:exrm, ">= 0.16.0", warn_missing: false},
        # {:distillery, ">= 0.8.0", optional: true, warn_missing: false},
      ]
```
And it works !
It seems that the "optional" lines are confusing elixir in my build environment, whereas they work in my dev environment.

**Comment by schnittchen** [comment]
That did not work in my case. On the build host, `mix deps.get` fetches exrm, but `mix deps.tree` prints distillery, not exrm.

**Comment by schnittchen** [comment]
Uh. It's probably because I was running `mix deps.get` and checking in `mix.lock` changes from the top umbrella app. Running the task from the project root produces the exrm dependency, since there is no rel/config.exs at the top.

**Comment by bharendt** [comment]
exrm and distillery cannot be used as dependency together because they both implement modules with the same name for the `mix release` and `mix release.clean` tasks. In addtion edeliver provides release plugins for both release tools (for linking configs and automatic relup modifications) which have compile-time dependencies to either one or the other. edeliver tries to autodetect which one is used, but unfortunately the check is not executed during compile time of the project which should be built with edeliver, but when the hex package is built, which means that exrm is always marked as required (default) and distillery is marked as optional. To fix this I marked both dependencies as `optional` (in the `distillery` branch) but unfortunately this change will not be backwards compatible because it requires to add the build tool (especially exrm which is the default) as a dependency in all existing elixir projects which are built and deployed using edeliver.

**Comment by Moxide** [comment]
Hello,
the last commit fixed my issues ;-)
Thanks for the solution, and the explanations.

**Comment by bharendt** [comment]
@Moxide: Thanks for the feedback. That changes are included in the [release 1.4.0](https://hex.pm/packages/edeliver/1.4.0).
